docs/source/howto/codes.rst

```diff
@@ -3,173 +3,200 @@
+    There are cases (e.g. license issues, file size) where you may want to change this behavior and :ref:`exclude files from being stored<topics:calculations:usage:calcjobs:file_lists_provenance_exclude>`.
 
-.. code-block:: python
+After having written the necessary input files, we let AiiDA know how to run the code via the |CodeInfo| object.
 
-    spec.input('code', valid_type=orm.Code, help='The `Code` to use for this job.')
+First, we forward the ``uuid`` of the |Code|  instance passed by the user via the generic ``code`` input mentioned previously (in this example, the ``code`` will represent a ``bash`` executable).
 
-After defining the UUID of the code node that the engine should use, we define the filenames where the stdin and stdout file descriptors should be redirected to.
-These values are taken from the inputs, which are part of the ``metadata.options`` namespace, for some of whose inputs we overrode the default values in the specification definition in the previous section.
-Note that instead of retrieving them through ``self.inputs.metadata['options']['input_filename']``, one can use the shortcut ``self.options.input_filename`` as we do here.
-Based on this definition of the |CodeInfo|, the engine will create a run script that looks like the following:
+Second, let's recall how we want our executable to be run:
 
 .. code-block:: bash
 
     #!/bin/bash
 
     '[executable path in code node]' < '[input_filename]' > '[output_filename]'
 
-The |CodeInfo| should be attached to the ``codes_info`` attribute of a |CalcInfo| object.
-A calculation can potentially run more than one code, so the |CodeInfo| object should be assigned as a list.
-Finally, we define the ``retrieve_list`` attribute, which is a list of filenames that the engine should retrieve from the running directory once the calculation job has finished.
+We want to pass our input file to the executable via standard input, and record standard output of the executable in the output file -- this is done using the ``stdin_name`` and ``stdout_name`` attributes of the |CodeInfo|.
+
+.. tip::
+
+     Many executables don't read from standard input but require the path to an input file to be passed via command line parameters (potentially including further configuration options) instead.
+     In that case, use the |CodeInfo| ``cmdline_params`` attribute:
+
+     .. code-block:: python
+
+         codeinfo.cmdline_params = [ '--input', self.inputs.input_filename ]
+
+.. tip::
+
+   The dot-notation ``self.options.input_filename`` and the key-notation ``self.options['input_filename']`` are equivalent for *accessing* the value of an option, but the key-notation provides more control when *setting* values (e.g. overriding the default value while preserving the ``valid_type``).
+   That is why we used the key-notation in the ``define`` method.
```

Comment:
I might have misunderstood what you are trying to say here, but the second part is incorrect. Note that the return type of `self.options` is an attribute dictionary of actual option values passed to the process and is completely different from ``spec.inputs['metadata']['options']`` which is a `PortNamespace`. The reason that we use key-notation in define is simply because the `PortNamespace` is not an attribute dict, unlike the return value of `self.options`. Note that there is *nothing to set* also in `prepare_for_submission`.

Reply:
Right, I now checked the old version of this and it just mentioned a short-hand notation; not the detail about ports etc. (which indeed should be explained in the `define` section, if any).
I will fix this.